src/core/symbology/qgssymbollayerutils.h

```diff
@@ -648,6 +648,36 @@ class CORE_EXPORT QgsSymbolLayerUtils
     //! Calculate whether a point is within of a QPolygonF
     static bool pointInPolygon( const QPolygonF &points, QPointF point );
 
+    /**
+     * Returns the substring of a \a polyline which starts at \a startOffset from the beginning of the line
+     * and ends at \a endOffset from the start of the line.
+     *
+     * If \a startOffset is less than 0, then the start point will be calculated by subtracting that distance
+     * from the end of the line. Similarly, if \a endOffset is less than zero then the end point will be subtracted
+     * from the end of the line.
+     *
+     * May return an empty linestring if the substring is zero length.
+     *
+     * \since QGIS 3.16
+     */
+    static QPolygonF polylineSubstring( const QPolygonF &polyline, double startOffset, double endOffset );
+
+    /**
+     * Returns TRUE if the angle formed by the line \a p1 - \a p2 - \a p3 forms a "sharp" corner.
```

Comment:
Might be worth adding in the dox what minimum angle is considered to be a "sharp" corner.